h2o-py/h2o/automl/_base.py

```diff
@@ -125,3 +125,76 @@ def event_log(self):
+                            - Multinomial metrics: mean_per_class_error, logloss, rmse, mse, auc, aucpr
+                          The following additional leaderboard information can be also used as a criterion:
+                            - 'training_time_ms': column providing the training time of each model in milliseconds (doesn't include the training of cross validation models).
+                            - 'predict_time_per_row_ms`: column providing the average prediction time by the model for a single row.
+        :return: a model or None if none of a given family is present
+        :examples:
+        >>> # Set up an H2OAutoML object
+        >>> aml = H2OAutoML(max_runtime_secs=30)
+        >>> # Launch an AutoML run
+        >>> aml.train(y=y, training_frame=train)
+        >>> gbm = aml.get_best_model("gbm")
+        """
+        patterns = dict(
+            basemodel="^(?!StackedEnsemble)",
+            deeplearning="^DeepLearning$",
+            drf="^DRF$",
+            gbm="^GBM$",
+            glm="^GLM$",
+            stackedensemble="^StackedEnsemble$",
+            xgboost="^XGBoost$",
+            any=".*"
+        )
+
+        higher_is_better = ["auc", "aucpr"]
+        model_category = self.leader._model_json["output"]["model_category"]
+        if "Regression" == model_category:
+            default_criterion = "mean_residual_deviance"
+        elif "Binomial" == model_category:
+            default_criterion = "auc"
+        else:
+            default_criterion = "mean_per_class_error"
+
+        if criterion is None:
+            criterion = [default_criterion]
+        else:
+            # Deal with potential ties when not using the default criterion => use it to break the ties
+            criterion = [criterion.lower(), default_criterion]
+
+        extra_cols = ["algo"]
+        if criterion[0] in ("training_time_ms", "predict_time_per_row_ms"):
+            extra_cols.append(criterion[0])
+
+        leaderboard = h2o.automl.get_leaderboard(self, extra_columns=extra_cols)
+        criteria = {col.lower(): col for col in leaderboard.columns}
+
+        if criterion[0].lower() not in criteria.keys():
+            from h2o.exceptions import H2OValueError
+            raise H2OValueError("Criterion \"{}\" is not present in the leaderboard!".format(criterion))
+
+        criterion = [criteria[c] for c in criterion]
+
+        if algorithm.lower() not in patterns.keys():
+            from h2o.exceptions import H2OValueError
+            raise H2OValueError("Incorrect model_type specified \"{}\". Has to be one of \"{}\"".format(
+                algorithm,
+                '", "'.join(sorted(list(patterns.keys())))
+            ))
+
+        sorted_lb = leaderboard.sort(by=criterion, ascending=[c.lower() not in higher_is_better for c in criterion])
+        matches = sorted_lb[int(sorted_lb["algo"].grep(patterns[algorithm.lower()])[0, :]), :]["model_id"]
```

Comment:
looks like this could be simplified now, without the pattern logic.
Also to solve the ties without replicating the backend logic, I suggest the following:
- filter the extended leaderboard by algo+criterion and sort the result by criterion
- collect all the model ids from the top as soon as they have the same criterion value as the top one.
- if more than one model was collected, pick the first model in default leaderboard present in the collected ones.

Reply:
Merci beaucoup! After you mentioned ties I went digging into the Merge - Radix sort we have implemented in h2o to find out if it is stable - I didn't find definitive answer but it doesn't seem to be. After that I completely overlooked this simple solution.